### PR TITLE
Add triplet hard loss with soft margin

### PR DESCRIPTION
**Pull request by HauserA**
Added batch hard loss with soft margin version as in [In Defense of the Triplet Loss for Person Re-Identification](https://arxiv.org/abs/1703.07737).

The difference to triplet_semihard_loss is the hard negative and hard positive mining and the option for a soft margin with the softplus-function instead of a hard margin.

### CODE REVIEW

**Comment by HauserA** [comment]
@googlebot I signed it!

File: tensorflow_addons/losses/triplet.py [hunk 1]
```diff
@@ -175,3 +227,43 @@ def get_config(self):
...
+        super().__init__(name=name, reduction=tf.keras.losses.Reduction.NONE)
+        self.margin = margin
+        self.soft = soft
+
+    def call(self, y_true, y_pred):
+        return triplet_hard_loss(y_true, y_pred, self.margin, self.soft)
+
+    def get_config(self):
+        config = {
+            "margin": self.margin,
```
Comment (WindQAQ) [inline tensorflow_addons/losses/triplet.py]: Should we also serialize `self.soft` here?

Comment (HauserA) [reply tensorflow_addons/losses/triplet.py]: Yes
